README.md
# DAMH1-SlangWord
Slang Word APP
Các chức năng
1. Chức năng tìm kiếm theo slang word.
2. Chức năng tìm kiếm theo definition, hiển thị ra tất cả các slang words mà trong
defintion có chứa keyword gõ vào.
3. Chức năng hiển thị history, danh sách các slang word đã tìm kiếm.
4. Chức năng add 1 slang words mới.
5. Chức năng edit 1 slang word.
6. Chức năng delete 1 slang word. Confirm trước khi xoá.
7. Chức năng reset danh sách slang words gốc.
8. Chức năng random 1 slang word (On this day slang word).
9. Chức năng đố vui, chương trình hiển thị 1 random slang word, với 4 đáp án cho
người dùng chọn.
10. Chức năng đố vui, chương trình hiển thị 1 definition, với 4 slang words đáp án cho
người dùng chọn.


history.txt
13/06/2023 16:27:56 |   $$
13/06/2023 16:28:17 |   $$
13/06/2023 16:29:30 |   $
13/06/2023 16:32:16 |   $
13/06/2023 16:38:06 |   $


src/test/SlangWordApp.java
package test;

import java.awt.EventQueue;
import javax.swing.UIManager;
import javax.swing.UIManager.LookAndFeelInfo;

import ui.SlangApp;
public class SlangWordApp {

	public static void main(String[] args) {
		EventQueue.invokeLater(new Runnable() {
			
			@Override
			public void run() {
				try {
				    for (LookAndFeelInfo info : UIManager.getInstalledLookAndFeels()) {
				        if ("Nimbus".equals(info.getName())) {
				            UIManager.setLookAndFeel(info.getClassName());
				            break;
				        }
				    }
				} catch (Exception e) {
				}
				new SlangApp();
			}
		});
	}

}


src/model/Dict.java
package model;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.File;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Iterator;
import java.util.LinkedHashMap;
import java.util.LinkedList;
import java.util.Map;
import java.util.Map.Entry;
import java.util.Random;

import java.io.*;
import java.util.*;

public class Dict {
    private HashMap<String, LinkedList<String>> dictionary;
    private LinkedList<String> historyList, deleteList;
    private LinkedHashMap<String, LinkedList<String>> modifiedList;

    public Dict() {
        dictionary = new HashMap<>();
        historyList = new LinkedList<>();
        deleteList = new LinkedList<>();
        modifiedList = new LinkedHashMap<>();

        loadMap(dictionary, "slang.txt");
        loadLinkedList(historyList, "history.txt");
        loadLinkedList(deleteList, "delete.txt");

        for (String slang : deleteList) {
            dictionary.remove(slang);
        }

        loadMap(modifiedList, "modified.txt");

        for (Map.Entry<String, LinkedList<String>> entry : modifiedList.entrySet()) {
            String key = entry.getKey();
            LinkedList<String> val = entry.getValue();
            dictionary.put(key, val);
        }
    }

    private void loadMap(Map<String, LinkedList<String>> map, String filename) {
        try (BufferedReader br = new BufferedReader(new FileReader(filename))) {
            String line;
            while ((line = br.readLine()) != null) {
                String[] parts = line.split("`");
                if (parts.length == 2) {
                    String slang = parts[0].trim();
                    String[] definitions = parts[1].split("\\|");
                    LinkedList<String> definitionList = new LinkedList<>(Arrays.asList(definitions));
                    map.put(slang, definitionList);
                }
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    private void loadLinkedList(LinkedList<String> list, String filename) {
        try (BufferedReader br = new BufferedReader(new FileReader(filename))) {
            String line;
            while ((line = br.readLine()) != null) {
                list.add(line);
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    private void saveLinkedList(LinkedList<String> list, String filename) {
        try (BufferedWriter bw = new BufferedWriter(new FileWriter(filename))) {
            for (String item : list) {
                bw.write(item);
                bw.newLine();
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    private void saveModified() {
        try (BufferedWriter bw = new BufferedWriter(new FileWriter("modified.txt"))) {
            for (Map.Entry<String, LinkedList<String>> entry : modifiedList.entrySet()) {
                String key = entry.getKey();
                LinkedList<String> val = entry.getValue();
                StringBuilder sb = new StringBuilder();
                for (String definition : val) {
                    sb.append(definition).append("|");
                }
                sb.deleteCharAt(sb.length() - 1);
                bw.write(key + "`" + sb.toString());
                bw.newLine();
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public void save() {
        saveLinkedList(deleteList, "delete.txt");
        saveLinkedList(historyList, "history.txt");
        saveModified();
    }

    public void addHistory(String history) {
        historyList.add(history);
    }

    public LinkedList<String> getHistory() {
        return historyList;
    }

    public LinkedList<String> searchSlang(String slang) {
        return dictionary.get(slang);
    }

    public void addDefinition(String slang, String definition) {
        LinkedList<String> definitions = dictionary.getOrDefault(slang, new LinkedList<>());
        definitions.add(definition);
        dictionary.put(slang, definitions);
        modifiedList.put(slang, definitions);
    }

    public void addNew(String slang, String definition) {
        LinkedList<String> definitions = new LinkedList<>();
        definitions.add(definition);
        dictionary.put(slang, definitions);
        modifiedList.put(slang, definitions);
    }

    public boolean hasSlang(String slang) {
        return dictionary.containsKey(slang);
    }

    public boolean editSlang(String slang, String oldDefinition, String newDefinition) {
        LinkedList<String> definitions = dictionary.get(slang);
        if (definitions != null) {
            int index = definitions.indexOf(oldDefinition);
            if (index != -1) {
                definitions.set(index, newDefinition);
                modifiedList.put(slang, definitions);
                return true;
            }
        }
        return false;
    }

    public void deleteSlang(String slang) {
        deleteList.add(slang);
        dictionary.remove(slang);
    }

    public ArrayList<String> searchDefinition(String definition) {
        ArrayList<String> slangs = new ArrayList<>();
        for (Map.Entry<String, LinkedList<String>> entry : dictionary.entrySet()) {
            String slang = entry.getKey();
            LinkedList<String> definitions = entry.getValue();
            for (String def : definitions) {
                if (def.contains(definition)) {
                    slangs.add(slang);
                    break;
                }
            }
        }
        return slangs;
    }

    public String randomSlang() {
        Random random = new Random();
        ArrayList<String> slangs = new ArrayList<>(dictionary.keySet());
        return slangs.get(random.nextInt(slangs.size()));
    }

    public HashMap<String, LinkedList<String>> slangGame() {
        HashMap<String, LinkedList<String>> game = new HashMap<>();
        Random random = new Random();
        ArrayList<String> slangs = new ArrayList<>(dictionary.keySet());
        while (game.size() < 4) {
            String slang = slangs.get(random.nextInt(slangs.size()));
            game.put(slang, dictionary.get(slang));
        }
        return game;
    }

    public HashMap<String, ArrayList<String>> definitionGame() {
        HashMap<String, ArrayList<String>> game = new HashMap<>();
        Random random = new Random();
        ArrayList<String> slangs = new ArrayList<>(dictionary.keySet());
        while (game.size() < 4) {
            String slang = slangs.get(random.nextInt(slangs.size()));
            LinkedList<String> definitions = dictionary.get(slang);
            String definition = definitions.get(random.nextInt(definitions.size()));
            ArrayList<String> slangsContainingDef = searchDefinition(definition);
            game.put(definition, slangsContainingDef);
        }
        return game;
    }

    public void reset() {
        dictionary.clear();
        modifiedList.clear();
        historyList.clear();
        deleteList.clear();
        loadMap(dictionary, "slang.txt");
        loadMap(modifiedList, "modified.txt");
    }
}


src/ui/SlangApp.java
package ui;

import java.awt.BorderLayout;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.awt.event.WindowAdapter;
import javax.swing.JFrame;
import javax.swing.JMenu;
import javax.swing.JMenuBar;
import javax.swing.JMenuItem;
import javax.swing.JPanel;
import javax.swing.JTabbedPane;

import model.Dict;

public class SlangApp implements ActionListener{
	
	public SlangApp() {
		dict = new Dict();
		frame = new JFrame("Slang app");
		frame.setDefaultCloseOperation(frame.EXIT_ON_CLOSE);
		frame.setLayout(new BorderLayout());
		frame.addWindowListener(new WindowAdapter() {
		    @Override
		    public void windowClosing(java.awt.event.WindowEvent windowEvent) {
		        dict.save();
		    }
		});
		
		JMenuBar menuBar = new JMenuBar();
		JMenu fileMenu = new JMenu("Clear History");
		
		reset = new JMenuItem("Reset Dictionary");
		reset.addActionListener(this);
		fileMenu.add(reset);
		menuBar.add(fileMenu);
		
		slangPanel = new SlangForm(dict);
		tabbedPane = new JTabbedPane();
		tabbedPane.addTab("Slang Form", slangPanel);
		tabbedPane.addTab("Search definition", new SearchForm(dict));
		tabbedPane.addTab("On this day", new RandomForm(dict));
        tabbedPane.addTab("Game Form", new SlangGame(dict));
        
		
		
		frame.setJMenuBar(menuBar);
		frame.add(tabbedPane, BorderLayout.CENTER);
		frame.setVisible(true);
		frame.pack();
	}
	@Override
	public void actionPerformed(ActionEvent e) {
		if (e.getSource() == reset) {
			System.out.println("Here");
			dict.reset();
			slangPanel.clearHistory();
		}
		
	}
    JFrame frame;
	JTabbedPane tabbedPane;
	SlangForm slangPanel;
	JPanel findDefiPanel;
	JMenuItem reset;
	Dict dict;
}

src/ui/SlangGame.java
package ui;

import java.awt.Color;
import java.awt.Font;
import java.awt.GridBagConstraints;
import java.awt.GridBagLayout;
import java.awt.Insets;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.Map.Entry;
import java.util.Random;

import javax.swing.BoxLayout;
import javax.swing.ButtonGroup;
import javax.swing.DefaultListModel;
import javax.swing.JButton;
import javax.swing.JLabel;
import javax.swing.JPanel;
import javax.swing.JRadioButton;

import model.Dict;

public class SlangGame extends JPanel implements ActionListener {
    JLabel word, messageLabel, solution;
    Dict dict;
    DefaultListModel<String> listModel;
    JButton slangGame;
    JPanel createPanel;
    JRadioButton A, B, C, D;
    ButtonGroup answerGroup;
    String key = "";
    boolean isSlang = false;

    public SlangGame(Dict d) {
        dict = d;
        addControls();
    }
    private void addControls() {
        
        setLayout(new GridBagLayout());
        GridBagConstraints c = new GridBagConstraints();
        c.insets = new Insets(5, 5, 5, 5);
        c.gridx = 0;
        c.gridy = 0;

        slangGame = new JButton("Create new slang game");
        slangGame.addActionListener(this);
        createPanel = new JPanel();
        createPanel.setLayout(new BoxLayout(createPanel, BoxLayout.X_AXIS));
        createPanel.add(slangGame);

        word = new JLabel();
        JPanel answerPanel = new JPanel();
        answerPanel.setLayout(new BoxLayout(answerPanel, BoxLayout.Y_AXIS));
        answerGroup = new ButtonGroup();
        A = new JRadioButton();
        B = new JRadioButton();
        C = new JRadioButton();
        D = new JRadioButton();
        A.addActionListener(this);
        B.addActionListener(this);
        C.addActionListener(this);
        D.addActionListener(this);
        answerGroup.add(A);
        answerGroup.add(B);
        answerGroup.add(C);
        answerGroup.add(D);

        answerPanel.add(A);
        answerPanel.add(B);
        answerPanel.add(C);
        answerPanel.add(D);

        word = new JLabel("", JLabel.CENTER);
        word.setFont(new Font("Verdana", Font.BOLD, 20));
        messageLabel = new JLabel("", JLabel.CENTER);
        messageLabel.setFont(new Font("Verdana", Font.ITALIC, 20));
        solution = new JLabel("", JLabel.CENTER);
        solution.setFont(new Font("Verdana", Font.BOLD, 25));

        add(createPanel, c);
        c.gridy = 1;
        add(word, c);
        c.gridy = 2;
        add(answerPanel, c);
        c.gridy = 3;
        add(messageLabel, c);
        c.gridy = 4;
        add(solution, c);
    }
    
    @Override
    public void actionPerformed(ActionEvent e) {
        A.setEnabled(true);
        B.setEnabled(true);
        C.setEnabled(true);
        D.setEnabled(true);
        solution.setText("");

        if (e.getSource() == slangGame) {
            startSlangGame();
        }

        if (A.isSelected()) {
            checkAnswer(A);
        }
        if (B.isSelected()) {
            checkAnswer(B);
        }
        if (C.isSelected()) {
            checkAnswer(C);
        }
        if (D.isSelected()) {
            checkAnswer(D);
        }
    }

    private void startSlangGame() {
        answerGroup.clearSelection();
        isSlang = true;
        messageLabel.setText("");
        Random r = new Random();
        HashMap<String, LinkedList<String>> myGame = dict.slangGame();
        ArrayList<String> slangs = new ArrayList<String>(myGame.keySet());
        key = slangs.get(r.nextInt(slangs.size()));
        word.setText("Slang: " + key);
        int count = 1;
        for (Entry<String, LinkedList<String>> entry : myGame.entrySet()) {
            LinkedList<String> value = entry.getValue();
            setAnswerOption(count, value.iterator().next());
            count++;
        }
    }

    private void setAnswerOption(int count, String value) {
        switch (count) {
            case 1:
                A.setText(value);
                break;
            case 2:
                B.setText(value);
                break;
            case 3:
                C.setText(value);
                break;
            case 4:
                D.setText(value);
                break;
        }
    }

    private void checkAnswer(JRadioButton selectedButton) {
        A.setEnabled(false);
        B.setEnabled(false);
        C.setEnabled(false);
        D.setEnabled(false);
        messageLabel.setText("");
        boolean isCorrect = false;
        String sol = "";

        if (isSlang) {
            sol = dict.searchSlang(key).iterator().next();
            if (selectedButton.getText() == sol) {
                isCorrect = true;
            }
        }

        if (isCorrect) {
            messageLabel.setText("Correct");
            messageLabel.setForeground(new Color(0, 102, 0));
        } else {
            messageLabel.setForeground(Color.RED);
            messageLabel.setText("Incorrect");
            solution.setText("Answer: " + sol);
            solution.setForeground(new Color(0, 102, 0));
        }
    }
}


src/ui/RandomForm.java
package ui;

import java.awt.GridBagConstraints;
import java.awt.GridBagLayout;
import java.awt.Insets;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.util.Iterator;
import javax.swing.JButton;
import javax.swing.JLabel;
import javax.swing.JPanel;

import model.Dict;

public class RandomForm extends JPanel {
	
	public RandomForm(Dict d) {
		dict = d;
		setLayout(new GridBagLayout());
		getWord = new JButton("Click!");
		label = new JLabel("On this day slang word", JLabel.CENTER);
		slang = new JLabel();
		definition = new JLabel();
		
		GridBagConstraints c = new GridBagConstraints ();
		c.insets = new Insets (10, 0, 0, 0);
		add(getWord, c);
		c.gridy = 1;
        add(label, c);
        c.gridy = 2;
        add(slang, c);
        c.gridy++;
        add(definition, c);
        
		getWord.addActionListener(new ActionListener() {
			
			@Override
			public void actionPerformed(ActionEvent e) {
				slang.setText(dict.randomSlang());
				String defs = "";
				for (Iterator iterator = dict.searchSlang(slang.getText()).iterator(); iterator.hasNext();) {
					String str = (String) iterator.next();
					defs += str + ", ";
				}
				definition.setText(defs.substring(0, defs.length()-2));
			}
		});
	}
    JButton getWord;
	JLabel label;
	JLabel slang;
	Dict dict;
	JLabel definition;
}

src/ui/SearchForm.java
package ui;

import java.awt.BorderLayout;
import java.awt.Color;
import java.awt.Dimension;
import java.awt.FlowLayout;
import java.awt.Font;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.LinkedList;

import javax.swing.BorderFactory;
import javax.swing.BoxLayout;
import javax.swing.DefaultListModel;
import javax.swing.JButton;
import javax.swing.JLabel;
import javax.swing.JList;
import javax.swing.JOptionPane;
import javax.swing.JPanel;
import javax.swing.JScrollPane;
import javax.swing.JSplitPane;
import javax.swing.JTextArea;
import javax.swing.JTextField;
import javax.swing.border.Border;
import javax.swing.border.TitledBorder;
import javax.swing.event.ListSelectionEvent;

import model.Dict;

public class SearchForm extends JPanel implements ActionListener {
	JButton searchButton;
	JList resultList;
	DefaultListModel<String> listModel;
	Dict dict;
	JLabel slangLabel, definitionLabel, searchDefinitionLabel;
	JTextField slangField, definitionField, searchDefinitionField;
	JScrollPane resultPane, historyPanel;
	JTextArea historyArea;

	public SearchForm(Dict d) {
		dict = d;
		addControls();
		addEvents();
	}

	private void addControls() {
		
		setLayout(new BorderLayout());

		JPanel pnLeft = new JPanel();
		pnLeft.setLayout(new BoxLayout(pnLeft, BoxLayout.Y_AXIS));
		pnLeft.setPreferredSize(new Dimension(300, 0));
		JPanel pnRight = new JPanel();
		JSplitPane sp = new JSplitPane(JSplitPane.HORIZONTAL_SPLIT, pnLeft, pnRight);
		pnRight.setLayout(new BoxLayout(pnRight, BoxLayout.Y_AXIS));
		add(sp, BorderLayout.CENTER);

		/////// TOP///////////
		JPanel pnTop = new JPanel();
		pnTop.setLayout(new FlowLayout());
		JLabel lblTop = new JLabel("Tu Dien Slang Word - 21424073");
		lblTop.setFont(new Font("Verdana", Font.PLAIN, 18));
		pnTop.add(lblTop);
		add(pnTop, BorderLayout.NORTH);

		//// Left//////////

		Border borderCRUD = BorderFactory.createLineBorder(Color.BLUE);
		TitledBorder borderTitleCRUD = new TitledBorder(borderCRUD, "CRUD APP");
		borderTitleCRUD.setTitleJustification(TitledBorder.CENTER);
		borderTitleCRUD.setTitleColor(Color.RED);
		pnLeft.setBorder(borderTitleCRUD);

		JPanel pnMa = new JPanel();
		pnMa.setLayout(new FlowLayout());
		slangLabel = new JLabel("Definition");
		slangField = new JTextField(15);
		pnMa.add(slangLabel);
		pnMa.add(slangField);
		pnLeft.add(pnMa);

		JPanel pnDef = new JPanel();
		pnDef.setLayout(new FlowLayout());
		definitionLabel = new JLabel("SlangWord");
		definitionField = new JTextField(15);
		pnDef.add(definitionLabel);
		pnDef.add(definitionField);
		pnLeft.add(pnDef);


		/////// Right/////////////
		JPanel pnSlang = new JPanel();

		Border borderSlang = BorderFactory.createLineBorder(Color.BLUE);
		TitledBorder borderTitleSlang = new TitledBorder(borderSlang, "Search Definition");
		borderTitleSlang.setTitleJustification(TitledBorder.CENTER);
		borderTitleSlang.setTitleColor(Color.RED);
		pnRight.setBorder(borderTitleSlang);

		pnSlang.setLayout(new FlowLayout());
		searchDefinitionLabel = new JLabel("Definition");
		searchDefinitionField = new JTextField(15);
		searchButton = new JButton("Search");
		searchButton.addActionListener(this);
		pnSlang.add(searchDefinitionLabel);
		pnSlang.add(searchDefinitionField);
		pnSlang.add(searchButton);
		pnRight.add(pnSlang);

		listModel = new DefaultListModel<String>();
		
		resultList = new JList(listModel);
		resultList.addListSelectionListener((ListSelectionEvent e) -> {

			definitionField.setText((String) resultList.getSelectedValue());
			slangField.setText(searchDefinitionField.getText());
		});
		resultPane = new JScrollPane(resultList, JScrollPane.VERTICAL_SCROLLBAR_ALWAYS,
				JScrollPane.HORIZONTAL_SCROLLBAR_ALWAYS);
		pnRight.add(resultPane);

		//////// BOTTOM//////////
		JPanel pnHistory = new JPanel();

		Border borderHistory = BorderFactory.createLineBorder(Color.RED);
		TitledBorder borderTitleHistory = new TitledBorder(borderHistory, "Lich su tim kiem");
		borderTitleHistory.setTitleJustification(TitledBorder.CENTER);
		borderTitleHistory.setTitleColor(Color.BLUE);
		pnHistory.setBorder(borderTitleHistory);

		pnHistory.setLayout(new BoxLayout(pnHistory, BoxLayout.Y_AXIS));
		historyArea = new JTextArea();
		historyArea.setEditable(false);
		historyPanel = new JScrollPane(historyArea, JScrollPane.VERTICAL_SCROLLBAR_ALWAYS,
				JScrollPane.HORIZONTAL_SCROLLBAR_ALWAYS);
		historyPanel.setPreferredSize(new Dimension(0, 150));

		pnHistory.add(historyPanel);

		LinkedList<String> history = dict.getHistory();
		for (String string : history) {
			historyArea.append("\n" + string);
		}
		add(pnHistory, BorderLayout.SOUTH);

		slangLabel.setPreferredSize(definitionLabel.getPreferredSize());
	}

	private void addEvents() {
		searchButton.addActionListener(new ActionListener() {

			@Override
			public void actionPerformed(ActionEvent e) {
				listModel.removeAllElements();
				ArrayList<String> slangList = dict.searchDefinition(searchDefinitionField.getText());
				SimpleDateFormat formatter = new SimpleDateFormat("dd/MM/yyyy HH:mm:ss");
				Date date = new Date();
				String time = formatter.format(date);
				String history = time + " |   " + searchDefinitionField.getText();
				if (slangList.size() != 0) {				
					for(String s : slangList) {
						listModel.addElement(s);
					}
				}
				else {
					JOptionPane.showMessageDialog(resultList, "Slang is not in dictionary");
				}
				dict.addHistory(history);
				historyArea.append("\n" + history);
			}
		});
	}

	public void clearHistory() {
		historyArea.setText("");
	}
	@Override
	public void actionPerformed(ActionEvent e) {
		// TODO Auto-generated method stub
		
	}
}


src/ui/SlangForm.java
package ui;

import java.awt.BorderLayout;
import java.awt.Color;
import java.awt.Dimension;
import java.awt.FlowLayout;
import java.awt.Font;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.HashSet;
import java.util.LinkedList;

import javax.swing.BorderFactory;
import javax.swing.BoxLayout;
import javax.swing.DefaultListModel;
import javax.swing.JButton;
import javax.swing.JLabel;
import javax.swing.JList;
import javax.swing.JOptionPane;
import javax.swing.JPanel;
import javax.swing.JScrollPane;
import javax.swing.JSplitPane;
import javax.swing.JTextArea;
import javax.swing.JTextField;
import javax.swing.border.Border;
import javax.swing.border.TitledBorder;
import javax.swing.event.ListSelectionEvent;

import model.Dict;

public class SlangForm extends JPanel{
	JButton addButton, editButton, clearButton, searchButton, deleteButton;
	JList resultList;
	DefaultListModel<String> listModel;
	Dict dict;
	JLabel slangLabel, definitionLabel, searchSlangLabel;
	JTextField slangField, definitionField, searchSlangField;
	JScrollPane resultPane, historyPanel;
	JTextArea historyArea;

	public SlangForm(Dict d) {
		dict = d;
		addControls();
		addEvents();
	}

	private void addControls() {
		
		setLayout(new BorderLayout());

		JPanel pnLeft = new JPanel();
		pnLeft.setLayout(new BoxLayout(pnLeft, BoxLayout.Y_AXIS));
		pnLeft.setPreferredSize(new Dimension(300, 0));
		JPanel pnRight = new JPanel();
		JSplitPane sp = new JSplitPane(JSplitPane.HORIZONTAL_SPLIT, pnLeft, pnRight);
		pnRight.setLayout(new BoxLayout(pnRight, BoxLayout.Y_AXIS));
		add(sp, BorderLayout.CENTER);

		/////// TOP///////////
		JPanel pnTop = new JPanel();
		pnTop.setLayout(new FlowLayout());
		JLabel lblTop = new JLabel("Tu Dien Slang Word - 21424073");
		lblTop.setFont(new Font("Verdana", Font.PLAIN, 18));
		pnTop.add(lblTop);
		add(pnTop, BorderLayout.NORTH);

		//// Left//////////

		Border borderCRUD = BorderFactory.createLineBorder(Color.BLUE);
		TitledBorder borderTitleCRUD = new TitledBorder(borderCRUD, "CRUD APP");
		borderTitleCRUD.setTitleJustification(TitledBorder.CENTER);
		borderTitleCRUD.setTitleColor(Color.RED);
		pnLeft.setBorder(borderTitleCRUD);

		JPanel pnMa = new JPanel();
		pnMa.setLayout(new FlowLayout());
		slangLabel = new JLabel("SlangWord");
		slangField = new JTextField(15);
		pnMa.add(slangLabel);
		pnMa.add(slangField);
		pnLeft.add(pnMa);

		JPanel pnDef = new JPanel();
		pnDef.setLayout(new FlowLayout());
		definitionLabel = new JLabel("Definition");
		definitionField = new JTextField(15);
		pnDef.add(definitionLabel);
		pnDef.add(definitionField);
		pnLeft.add(pnDef);

		JPanel pnLeftBtn = new JPanel();
		pnLeftBtn.setLayout(new FlowLayout());
		addButton = new JButton("Add");
		editButton = new JButton("Update");
		clearButton = new JButton("Clear");
		pnLeftBtn.add(addButton);
		pnLeftBtn.add(editButton);
		pnLeftBtn.add(clearButton);
		pnLeft.add(pnLeftBtn);

		/////// Right/////////////
		JPanel pnSlang = new JPanel();

		Border borderSlang = BorderFactory.createLineBorder(Color.BLUE);
		TitledBorder borderTitleSlang = new TitledBorder(borderSlang, "Search SlangWord");
		borderTitleSlang.setTitleJustification(TitledBorder.CENTER);
		borderTitleSlang.setTitleColor(Color.RED);
		pnRight.setBorder(borderTitleSlang);

		pnSlang.setLayout(new FlowLayout());
		searchSlangLabel = new JLabel("SlangWord");
		searchSlangField = new JTextField(15);
		searchButton = new JButton("Search");
		pnSlang.add(searchSlangLabel);
		pnSlang.add(searchSlangField);
		pnSlang.add(searchButton);
		pnRight.add(pnSlang);

		listModel = new DefaultListModel<>();
		resultList = new JList(listModel);
		resultList.addListSelectionListener((ListSelectionEvent e) -> {

			definitionField.setText((String) resultList.getSelectedValue());
			slangField.setText(searchSlangField.getText());
		});
		resultPane = new JScrollPane(resultList, JScrollPane.VERTICAL_SCROLLBAR_ALWAYS,
				JScrollPane.HORIZONTAL_SCROLLBAR_ALWAYS);
		deleteButton = new JButton("Delete");
		pnRight.add(resultPane);
		pnRight.add(deleteButton);

		//////// BOTTOM//////////
		JPanel pnHistory = new JPanel();

		Border borderHistory = BorderFactory.createLineBorder(Color.RED);
		TitledBorder borderTitleHistory = new TitledBorder(borderHistory, "Lich su tim kiem");
		borderTitleHistory.setTitleJustification(TitledBorder.CENTER);
		borderTitleHistory.setTitleColor(Color.BLUE);
		pnHistory.setBorder(borderTitleHistory);

		pnHistory.setLayout(new BoxLayout(pnHistory, BoxLayout.Y_AXIS));
		historyArea = new JTextArea();
		historyArea.setEditable(false);
		historyPanel = new JScrollPane(historyArea, JScrollPane.VERTICAL_SCROLLBAR_ALWAYS,
				JScrollPane.HORIZONTAL_SCROLLBAR_ALWAYS);
		historyPanel.setPreferredSize(new Dimension(0, 150));

		pnHistory.add(historyPanel);

		LinkedList<String> history = dict.getHistory();
		for (String string : history) {
			historyArea.append("\n" + string);
		}
		add(pnHistory, BorderLayout.SOUTH);

		definitionLabel.setPreferredSize(slangLabel.getPreferredSize());
	}

	private void addEvents() {
		addButton.addActionListener(new ActionListener() {

			@Override
			public void actionPerformed(ActionEvent e) {
				String slang = slangField.getText();
				String definition = definitionField.getText();
				if (dict.hasSlang(slang)) {
					String[] options = { "Tạo mới", "Ghi đè" };
					int choice = JOptionPane.showOptionDialog(null, "Slangword đã tồn tại", "Slangword đã tồn tại",
							JOptionPane.DEFAULT_OPTION, JOptionPane.WARNING_MESSAGE, null, options, options[0]);
					if (choice == 0) {
						dict.addDefinition(slang, definition);
						if (slang.equals(searchSlangField.getText())) {
							listModel.addElement(definition);
						}
					} else {
						dict.addNew(slang, definition);
						if (slang.equals(searchSlangField.getText())) {
							listModel.removeAllElements();
							listModel.addElement(definition);
						}
					}

				} else {
					dict.addNew(slang, definition);
					JOptionPane.showMessageDialog(null, "Thêm thành công");
				}

			}
		});
		editButton.addActionListener(new ActionListener() {

			@Override
			public void actionPerformed(ActionEvent e) {
				if (dict.editSlang(searchSlangField.getText(), (String) resultList.getSelectedValue(),
						definitionField.getText())) {
					listModel.addElement(definitionField.getText());
					listModel.removeElement(resultList.getSelectedValue());
				} else {
					JOptionPane.showMessageDialog(addButton, "Không tìm thấy slangword");
				}

			}
		});
		deleteButton.addActionListener(new ActionListener() {

			@Override
			public void actionPerformed(ActionEvent e) {
				int choice = JOptionPane.showConfirmDialog(deleteButton, "Slangword sẽ được xóa", "Xóa",
						JOptionPane.YES_NO_OPTION);
				if (choice == 0) {
					dict.deleteSlang(searchSlangField.getText());
					listModel.removeAllElements();
					slangField.setText("");
					searchSlangField.setText("");
				}

			}
		});
		searchButton.addActionListener(new ActionListener() {

			@Override
			public void actionPerformed(ActionEvent e) {
				listModel.removeAllElements();
				LinkedList<String> defSet = dict.searchSlang(searchSlangField.getText());
				SimpleDateFormat formatter = new SimpleDateFormat("dd/MM/yyyy HH:mm:ss");
				Date date = new Date();
				String time = formatter.format(date);
				String history = time + " |   " + searchSlangField.getText();
				if (defSet != null) {
					for (String s : defSet) {
						listModel.addElement(s.strip());
					}
				} else {
					JOptionPane.showMessageDialog(resultList, "Slangword is not in dictionary");
				}
				dict.addHistory(history);
				historyArea.append("\n" + history);

			}
		});
		clearButton.addActionListener(new ActionListener() {

			@Override
			public void actionPerformed(ActionEvent e) {
				slangField.setText("");
				definitionField.setText("");

			}
		});
	}

	public void clearHistory() {
		historyArea.setText("");
	}

}
